core/src/main/java/io/questdb/cairo/wal/WalUtils.java

```diff
@@ -96,4 +106,73 @@ public static void createTxnLogFile(FilesFacade ff, MemoryMARW mem, Path txnSeqD
+     * or a transaction number greater than -1 if it is a valid last refresh transaction
+     */
+    public static long getMatViewLastRefreshBaseTxn(
+            Path tablePath,
+            TableToken tableToken,
+            CairoConfiguration configuration,
+            MemoryCMR txnLogMemory,
+            WalEventReader walEventReader,
+            BlockFileReader blockFileReader,
+            MatViewStateReader matViewStateReader
+    ) {
+        long txnNotFound = -1;
+        long txnInvalid = -2;
+        try (MemoryCMR mem = txnLogMemory) {
+            int tablePathLen = tablePath.size();
+            mem.smallFile(configuration.getFilesFacade(), tablePath.concat(SEQ_DIR).concat(TXNLOG_FILE_NAME).$(), MemoryTag.MMAP_TX_LOG);
+            long txnCount = mem.getLong(TableTransactionLogFile.MAX_TXN_OFFSET_64);
+            if (txnCount > 0) {
+                long fileSize = TableTransactionLogFile.HEADER_SIZE + txnCount * TableTransactionLogV1.RECORD_SIZE;
+                if (mem.size() >= fileSize) {
+                    for (long txn = txnCount - 1; txn >= 0; txn--) {
+                        tablePath.trimTo(tablePathLen);
+                        long offset = TableTransactionLogFile.HEADER_SIZE + txn * TableTransactionLogV1.RECORD_SIZE;
+                        final int walId = mem.getInt(offset + TableTransactionLogFile.TX_LOG_WAL_ID_OFFSET);
+                        final int segmentId = mem.getInt(offset + TableTransactionLogFile.TX_LOG_SEGMENT_OFFSET);
+                        final int segmentTxn = mem.getInt(offset + TableTransactionLogFile.TX_LOG_SEGMENT_TXN_OFFSET);
+                        if (walId > 0) {
+                            tablePath.concat(WAL_NAME_BASE).put(walId).slash().put(segmentId);
+                            try (WalEventReader eventReader = walEventReader) {
+                                WalEventCursor walEventCursor = eventReader.of(tablePath, segmentTxn);
+                                if (walEventCursor.getType() == MAT_VIEW_DATA) {
+                                    return walEventCursor.getDataInfoExt().getLastRefreshBaseTableTxn();
+                                }
+                                if (walEventCursor.getType() == MAT_VIEW_INVALIDATE) {
+                                    return txnInvalid;
+                                }
+                            } catch (Throwable th) {
+                                // walEventReader may not be able to find/open the WAL-e files
+                                try (BlockFileReader blockReader = blockFileReader) {
+                                    tablePath.trimTo(tablePathLen).concat(MatViewState.MAT_VIEW_STATE_FILE_NAME);
+                                    blockReader.of(tablePath.$());
+                                    matViewStateReader.of(blockReader, tableToken);
+                                    // read from state file if exists
+                                    if (!matViewStateReader.isInvalid()) {
```

Comment:
According to the method's javadoc, in `else` branch here we should return -2, don't we?